pkg/util/tpm/tpm.go

```diff
@@ -0,0 +1,424 @@
+		return nil, err
+	}
+
+	eksha := sha1.Sum(ekcert)
+	ekhash := hex.EncodeToString(eksha[:])
+	tpm = &Tpm{}
+	unstructuredTpm := &runtime.Unstructured{}
+	unstructuredTpm, err = t.tpmclient.Get(ekhash)
+
+	if err != nil {
+		if allowEmpty == false {
+			return nil, fmt.Errorf("TPM does not exist and automatic creation is forbidden")
+		}
+		err = verification.VerifyEKCert(ekcert)
+		if err != nil {
+			return nil, err
+		}
+		tpm = &Tpm{
+			EKCert: ekcert,
+		}
+		unstructuredTpm.SetKind("Tpm")
+		unstructuredTpm.SetAPIVersion("coreos.com/v1")
+		unstructuredTpm.SetName(ekhash)
+		unstructuredTpm.Object = make(map[string]interface{})
+		unstructuredTpm.Object["EKCert"] = base64.StdEncoding.EncodeToString(tpm.EKCert)
+		unstructuredTpm.Object["AIKPub"] = ""
+		unstructuredTpm.Object["AIKBlob"] = ""
+
+		unstructuredTpm, err = t.tpmclient.Create(unstructuredTpm)
+		if err != nil {
+			return nil, err
+		}
+		unstructuredTpm, err = t.tpmclient.Get(ekhash)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tpm.EKCert, _ = base64.StdEncoding.DecodeString(unstructuredTpm.Object["EKCert"].(string))
+	tpm.AIKPub, _ = base64.StdEncoding.DecodeString(unstructuredTpm.Object["AIKPub"].(string))
+	tpm.AIKBlob, _ = base64.StdEncoding.DecodeString(unstructuredTpm.Object["AIKBlob"].(string))
+
+	if len(tpm.EKCert) == 0 {
+		tpm.EKCert = ekcert
+	}
+	if len(tpm.AIKPub) == 0 || len(tpm.AIKBlob) == 0 {
+		secret := make([]byte, 16)
+		_, err = rand.Read(secret)
+		if err != nil {
+			return nil, err
+		}
+		aikpub, aikblob, err := c.GenerateAIK()
+		if err != nil {
+			return nil, err
+		}
+		asymenc, symenc, err := verification.GenerateChallenge(ekcert, aikpub, secret)
+		if err != nil {
+			return nil, err
+		}
+		response, err := c.ValidateAIK(aikblob, asymenc, symenc)
+		if err != nil {
+			return nil, err
+		}
+		if !bytes.Equal(response[:], secret) {
+			return nil, fmt.Errorf("AIK could not be validated")
+		}
+		tpm.AIKPub = aikpub
+		tpm.AIKBlob = aikblob
+		unstructuredTpm.SetName(ekhash)
+		unstructuredTpm.Object["EKCert"] = tpm.EKCert
+		unstructuredTpm.Object["AIKPub"] = tpm.AIKPub
+		unstructuredTpm.Object["AIKBlob"] = tpm.AIKBlob
+		unstructuredTpm, err = t.tpmclient.Update(unstructuredTpm)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tpm.Address = address
+	return tpm, nil
+}
+
+func Quote(tpm *Tpm) ([][]byte, []tspiconst.Log, error) {
+	c := tpmclient.New(tpm.Address, 30*time.Second)
+	quote, log, err := c.GetQuote(tpm.AIKPub, tpm.AIKBlob, []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15})
+	return quote, log, err
+}
+
+func ValidateLogConsistency(log []tspiconst.Log) error {
+	for _, entry := range log {
+		hash := sha1.Sum(entry.Event[:])
+		if bytes.Equal(hash[:], entry.PcrValue[:]) {
+			continue
+		}
+		return fmt.Errorf("Log entry is inconsistent with claimed PCR value")
+	}
+
+	return nil
+}
+
+func ValidateLog(log []tspiconst.Log, quote [][]byte) error {
+	var virt_pcrs [24][20]byte
+
+	for _, entry := range log {
```

Comment:
Maybe just me, but might be worth documenting what is going on here (what each slice means, what the intention is)